stdlib/pqueue.mli

```diff
@@ -0,0 +1,208 @@
+        elements of array [a]. Runs in linear time. *)
+
+    val of_list: (prio * 'a) list -> 'a t
+    (** [of_list l] returns a new priority queue containing the
+        elements of list [l]. Runs in linear time. *)
+
+    val of_seq: (prio * 'a) Seq.t -> 'a t
+    (** [of_seq s] returns a new priority queue containing the
+        elements of sequence [s]. Runs in linear time. *)
+
+    (** {1:iteration Iteration}
+
+        The order in which the elements of a priority queue are
+        traversed is unspecified.
+
+        It is a programming error to mutate a priority queue (by
+        adding or removing elements) during an iteration of the queue.
+        Such an error may be detected and signaled by the backing dynamic
+        array implementation, but this is not guaranteed. *)
+
+    val iter: (prio -> 'a -> unit) -> 'a t -> unit
+    (** [iter f q] calls [f] on each element of [q] in some
+        unspecified order. *)
+
+    val fold: ('acc -> prio -> 'a -> 'acc) -> 'acc -> 'a t -> 'acc
+    (** [fold f accu q] is [(f (... (f (f accu x1) x2) ...) xn)]
+        where [x1,x2,...,xn] are the elements of [q] in some
+        unspecified order. *)
+
+    val to_seq: 'a t -> (prio * 'a) Seq.t
+    (** [to_seq q] is the sequence of elements of [q] in some
+        unspecified order. *)
+
+  end
+(** Output signature of the functor {!Make}. *)
+
+module Make (Prio : OrderedType) : S with type prio = Prio.t
+(** Functor building an implementation of the priority queue structure
+    given a totally ordered type. *)
+
+module MinQueue : S with type prio = int
+(** Min-priority queues with integer priorities. *)
+
+module MaxQueue : S with type prio = int
+(** Max-priority queues with integer priorities. *)
+
+module MakePoly (E : sig
+    type 'a t
+    val compare: 'a t -> 'a t -> int
+  end) : sig
+    type 'a elt = 'a E.t
+    type 'a t
+    val create: unit -> 'a t
+    val length: 'a t -> int
+    val is_empty: 'a t -> bool
+    val add: 'a t -> 'a elt -> unit
+    val add_seq: 'a t -> 'a elt Seq.t -> unit
+    exception Empty
+    val min_elt: 'a t -> 'a elt
+    val min_elt_opt: 'a t -> 'a elt option
+    val pop_min: 'a t -> 'a elt
+    val pop_min_opt: 'a t -> 'a elt option
+    val remove_min: 'a t -> unit
+    val clear: 'a t -> unit
+    val copy: 'a t -> 'a t
+    val of_array: 'a elt array -> 'a t
+    val of_list: 'a elt list -> 'a t
+    val of_seq: 'a elt Seq.t -> 'a t
+    val iter: ('a elt -> unit) -> 'a t -> unit
+    val fold: ('acc -> 'a elt -> 'acc) -> 'acc -> 'a t -> 'acc
+    val to_seq: 'a t -> 'a elt Seq.t
+  end
+(** Functor building an implementation of priority queues
+    given a totally ordered type of elements. This is convenient when
+    the type of elements is already equipped with a comparison function.
+    For instance, we get priority queues with string elements as follows:
+
+{[
+  module IntPQ = Pqueue.MakePoly(struct
+    type _ t = string
```

Comment:
Why `string` instead of `int`? (The intention of `IntPQ` was to suggest "priority queue of integers".)

Reply:
Oups, a last-minute change typo. I thought that `string` was more interesting than `int`, since `MinQueue` already provides queues of integers. I'll fix that.